// nblib/nblib.js

const axios = require('axios');

class NBLib{
	constructor(API,token){
		if(API){
			this.API = API;
		}else{
			this.API = "https://api.nbdomain.com/v1/"
		}
		if(token){
			this.token = token;
		}
	}
 async readDomain (domain){
 	const url = this.API+"?nid="+domain;
 	const myInit = {
            method: 'GET',
            headers: { 'content-type': 'application/json;charset=UTF-8', 'Authorization': 'Bearer '+this.token },
            mode: 'cors',
            cache: 'default'
        };
 	const res = await axios.get(url,myInit);
 	if(res){
 		return res.data;
 	}
 	return null;
 }
 async getAll(address){
 	const url = this.API+"query/";
 	const option = {
            headers: { 'content-type': 'application/json;charset=UTF-8', 'Authorization': 'Bearer '+this.token },
            mode: 'cors',
            cache: 'default'
        };
    const data = {
    	"k":"owner",
    	"v":address,
    	"t":"e"
    }
    console.log(url);
 	const res = await axios.post(url,data,option);
 	if(res){
 		return res.data;
 	}
 	return null;
 }

}

module.exports = NBLib;
